src/Workspaces/Core/MSBuild.BuildHost/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -391,132 +391,75 @@ public void RemoveDocument(string filePath)
-        }
-
-        private static string? s_frameworkRoot;
-        private static string FrameworkRoot
-        {
-            get
-            {
-                if (RoslynString.IsNullOrEmpty(s_frameworkRoot))
-                {
-                    var runtimeDir = System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory();
-                    s_frameworkRoot = Path.GetDirectoryName(runtimeDir); // back out one directory level to be root path of all framework versions
-                }
-
-                return s_frameworkRoot ?? throw new InvalidOperationException($"Unable to get {nameof(FrameworkRoot)}");
-            }
-        }
+            if (hintPath is not null)
+                metadata.Add(MetadataNames.HintPath, hintPath);
 
-        private static bool IsFrameworkReferenceAssembly(string filePath)
-        {
-            return PathUtilities.IsChildPath(FrameworkRoot, filePath);
+            _loadedProject.AddItem(ItemNames.Reference, metadataReferenceIdentity, metadata);
         }
 
-        public void RemoveMetadataReference(MetadataReference reference, AssemblyIdentity identity)
+        public void RemoveMetadataReference(string shortAssemblyName, string fullAssemblyName, string filePath)
         {
             if (_loadedProject is null)
             {
                 return;
             }
 
-            if (reference is PortableExecutableReference peRef && peRef.FilePath != null)
+            var item = FindReferenceItem(shortAssemblyName, fullAssemblyName, filePath);
+            if (item != null)
             {
-                var item = FindReferenceItem(identity, peRef.FilePath);
-                if (item != null)
-                {
-                    _loadedProject.RemoveItem(item);
-                }
+                _loadedProject.RemoveItem(item);
             }
         }
 
-        private MSB.Evaluation.ProjectItem FindReferenceItem(AssemblyIdentity identity, string filePath)
+        private MSB.Evaluation.ProjectItem FindReferenceItem(string shortAssemblyName, string fullAssemblyName, string filePath)
         {
-            if (_loadedProject is null)
-            {
-                throw new InvalidOperationException($"Unable to find reference item '{identity?.Name}'");
-            }
+            Contract.ThrowIfNull(_loadedProject, "The project was not loaded.");
 
             var references = _loadedProject.GetItems(ItemNames.Reference);
             MSB.Evaluation.ProjectItem? item = null;
 
             var fileName = Path.GetFileNameWithoutExtension(filePath);
 
-            if (identity != null)
-            {
-                var shortAssemblyName = identity.Name;
-                var fullAssemblyName = identity.GetDisplayName();
-
-                // check for short name match
-                item = references.FirstOrDefault(it => string.Compare(it.EvaluatedInclude, shortAssemblyName, StringComparison.OrdinalIgnoreCase) == 0);
+            // check for short name match
```

Comment:
i guess we had this logic before.  but it's reallly oogy to me that we're doing all this munging.

Reply:
Yeah it's not ideal at all, but changing it is well out of scope of this PR.